Add support for WFS layers

JahidAhmed opened the pull request:
## Description
Closes #2768 
Closes #2748 
Closes #2726 

Restructured loader service vs. file code by adding 'validation' / error checking for both services and files to ensure we don't need to manually differentiate between different types 

Allows WFS layers to be specified through the config as well

## Testing
WFS Layer used when testing (zoom to layer boundary to see layer): https://www.ldproxy.nrw.de/rest/services/kataster/collections/flurstueck/items

Sample 68 shows an example of WFS with limit defined in the config (there may be a long delay, where the legend and map are empty until our requests resolve... this still needs to be fixed with a loading indicator / progress bar of some sort)

Tested in Chrome and IE

## Documentation
JSDoc, In-line comments

## Checklist
<!-- Quick checklist for items that are easy to miss -->

- [x] Commit messages follow [the guidelines](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/master/CONTRIBUTING.md#-git-commit-guidelines)
- [x] Release notes have been updated
- [x] PR targets the correct release version
- [x] Help files and documentation have been updated

Remember, it is a *muffin offence* to open a PR with any of the above checklist items incomplete.

Please keep the original issue up to date with the final solution, expected behaviour, and any additional notes for testers

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763)
<!-- Reviewable:end -->

barryytm commented:
Review status: 18 of 19 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/polyfill/ie-polyfills.js, line 0 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LEKa7K51KfOBgZLtZlm:-LEKa7K51KfOBgZLtZln:b-92bll5) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/4751e6fbb7ebb2164cfcdf0ebe675cd885692543/src/polyfill/ie-polyfills.js#L0)):*
Was this added by accident?  If so, please remove it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763)*
<!-- Sent from Reviewable.io -->

dane-thomas commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-:-LEKcuku1jZRiHhZsKGT:bnfp4nl)*
<!-- Sent from Reviewable.io -->

dane-thomas commented:
Appears to be working in Firefox as well

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-:-LEKduqt6J14K9_m6TuG:b-vysha0)*
<!-- Sent from Reviewable.io -->

ShrutiVellanki commented:
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-:-LEKmgFrCNpzt-EfHsqP:b-109e86)*
<!-- Sent from Reviewable.io -->

james-rae commented:
Reviewed 2 of 8 files at r1, 17 of 17 files at r2.
Review status: all files reviewed, 4 unresolved discussions (waiting on @JahidAhmed and @ShrutiVellanki)

---

*[package.json, line 29 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LEysJdlBf5TfDLIOgK7:-LEysJdm96aEuuwY5MMD:b-f8qjn7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/4751e6fbb7ebb2164cfcdf0ebe675cd885692543/package.json#L29)):*
> ```JSON
>     "dotjem-angular-tree": "github:dotJEM/angular-tree",
>     "file-saver": "1.3.8",
>     "geoApi": "github:fgpv-vpgf/geoApi#v2.4.0-9",
> ```

this will bump to 2.5.0-1.  leaving this as blocking comment until v2.4.0 of ramp is released

---

*[src/app/core/config.class.js, line 139 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LEytCTtBlcLbqEA-WAg:-LEytCTtBlcLbqEA-WAh:b-jf7wd8) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/4751e6fbb7ebb2164cfcdf0ebe675cd885692543/src/app/core/config.class.js#L139)):*
> ```JavaScript
>                     boundingBox: false,
>                     query: true,
>                     snapshot: false,
> ```

technically, it is always in snapshot mode.  we may want to alternately consider not providing a snapshot item on the control panel, as it may confuse people (it would always be "on/disabled")

---

*[src/app/core/config.class.js, line 156 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LEytT4iA09WYv2V5OWA:-LEytT4iA09WYv2V5OWB:b91goux) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/4751e6fbb7ebb2164cfcdf0ebe675cd885692543/src/app/core/config.class.js#L156)):*
> ```JavaScript
>                     'data',
>                     'symbology',
>                     'interval'
> ```

not sure we have support for intervals. would need to be manually re-loaded on a timer.

---

*[src/app/geo/layer-blueprint.class.js, line 320 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LEz1WtrAm0bMEGzKgpX:-LEz1Wts4xvsiXlaMMql:blkyznf) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/4751e6fbb7ebb2164cfcdf0ebe675cd885692543/src/app/geo/layer-blueprint.class.js#L320)):*
> ```JavaScript
>      */
>     function _isFileOrWFSLayer(source) {
>         const fileTypes = [Geo.Service.Types.CSV, Geo.Service.Types.GeoJSON, Geo.Service.Types.Shapefile];
> ```

should WFS we in this array?

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763)*
<!-- Sent from Reviewable.io -->

JahidAhmed commented:
Review status: 9 of 22 files reviewed, 4 unresolved discussions (waiting on @barryytm, @dane-thomas, @james-rae, @JahidAhmed, and @ShrutiVellanki)

---

*[package.json, line 29 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LEysJdlBf5TfDLIOgK7:-LF2w84f07zD3pQySGi2:b-mhsq5s) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/4751e6fbb7ebb2164cfcdf0ebe675cd885692543/package.json#L29)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

this will bump to 2.5.0-1.  leaving this as blocking comment until v2.4.0 of ramp is released
</blockquote></details>

donethanks

---

*[src/app/core/config.class.js, line 139 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LEytCTtBlcLbqEA-WAg:-LF2w9fb28x6-17b2FZN:big5lma) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/4751e6fbb7ebb2164cfcdf0ebe675cd885692543/src/app/core/config.class.js#L139)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

technically, it is always in snapshot mode.  we may want to alternately consider not providing a snapshot item on the control panel, as it may confuse people (it would always be "on/disabled")
</blockquote></details>

removed it to avoid any confusion

---

*[src/app/core/config.class.js, line 156 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LEytT4iA09WYv2V5OWA:-LF2wBn4Dj4_Y9ONYAkp:bp5ywni) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/4751e6fbb7ebb2164cfcdf0ebe675cd885692543/src/app/core/config.class.js#L156)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

not sure we have support for intervals. would need to be manually re-loaded on a timer.
</blockquote></details>

removed thanks

---

*[src/app/geo/layer-blueprint.class.js, line 320 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LEz1WtrAm0bMEGzKgpX:-LF2wD9IB36iJ59rE7SQ:bhb2bi1) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/4751e6fbb7ebb2164cfcdf0ebe675cd885692543/src/app/geo/layer-blueprint.class.js#L320)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

should WFS we in this array?
</blockquote></details>

the type client-side is WFS but the underlying service type is still `GeoJSON` as its created through the same structure

---

*[src/polyfill/ie-polyfills.js, line  at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LEKa7K51KfOBgZLtZlm:-LF2wPgk0f0XDAl0W4QX:b-bcku0w) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/4751e6fbb7ebb2164cfcdf0ebe675cd885692543/src/polyfill/ie-polyfills.js#L)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Was this added by accident?  If so, please remove it.
</blockquote></details>

this was intentional, required for encoding information

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763)*
<!-- Sent from Reviewable.io -->

barryytm commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 4 files at r3, 11 of 11 files at r4.
Review status: all files reviewed, 4 unresolved discussions (waiting on @james-rae and @ShrutiVellanki)

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-:-LF2wl__BLttDXatnCNy:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ShrutiVellanki commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed, 4 unresolved discussions (waiting on @james-rae and @ShrutiVellanki)

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-:-LF2xVLe3KztHHdsndaU:bnfp4nl)*
<!-- Sent from Reviewable.io -->

dane-thomas commented:
Reviewed 2 of 4 files at r3, 11 of 11 files at r4.
Review status: all files reviewed, 4 unresolved discussions (waiting on @james-rae)

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763)*
<!-- Sent from Reviewable.io -->

james-rae commented:
Reviewed 2 of 4 files at r3, 8 of 11 files at r4, 1 of 2 files at r5, 1 of 2 files at r6, 1 of 1 files at r7.
Review status: all files reviewed, 3 unresolved discussions (waiting on @JahidAhmed)

---

*[schema.json, line 586 at r7](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LFce3Co3ufnS125U2Fq:-LFce3CpDGYftEUMg8_9:bns7csy) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/6dce2a4388fe7f4d2bfec5bd2ac44106b3673cc5/schema.json#L586)):*
> ```JSON
>                 "id": { "type": "string", "description": "The id of the layer for referencing within the viewer (does not relate directly to any external service)" },
>                 "name": { "type": "string", "description": "The display name of the layer.  If it is not present the viewer will make an attempt to scrape this information." },
>                 "nameField": { "type": "string", "description": "The display field of the layer.  If it is not present the viewer will make an attempt to scrape this information." },
> ```

Will it actually attempt to scrape?

---

*[src/app/geo/legend.service.js, line 134 at r7](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LFcfl0s9ueLsIVYaaev:-LFcfl0s9ueLsIVYaaew:bw0x7nn) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/6dce2a4388fe7f4d2bfec5bd2ac44106b3673cc5/src/app/geo/legend.service.js#L134)):*
> ```JavaScript
>      * @param {pos} optional position for layer to be on the legend
>      * @param {Boolean} addToLegend   indicates whether layer should be automatically added to legend. default true
>      * @returns {LegendBlock} returns a corresponding, newly created legend block
> ```

not returning anything anymore

---

*[src/app/ui/loader/layer-source.service.js, line 148 at r7](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LFchbt6DIe5t4Vhx30C:-LFchbt6DIe5t4Vhx30D:b-kxsrw4) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/6dce2a4388fe7f4d2bfec5bd2ac44106b3673cc5/src/app/ui/loader/layer-source.service.js#L148)):*
> ```JavaScript
>          * @return {Promise} a promsie resolving with the layer GeoJSON
>          */
>         function _getWFSData(serviceUrl, startIndex = 0, limit = 10, wfsResponse) { // TODO: look into removing startIndex and always going from 0 to the end
> ```

maybe limit = 200?  10 will cause a lot of server requests for any decent sized layer

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763)*
<!-- Sent from Reviewable.io -->

JahidAhmed commented:
Review status: 16 of 22 files reviewed, 3 unresolved discussions (waiting on @ShrutiVellanki, @barryytm, @dane-thomas, @james-rae, and @JahidAhmed)

---

*[schema.json, line 586 at r7](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LFce3Co3ufnS125U2Fq:-LFcoQQzDEqV9jT5ChC6:bx0fdab) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/6dce2a4388fe7f4d2bfec5bd2ac44106b3673cc5/schema.json#L586)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

Will it actually attempt to scrape?
</blockquote></details>

it did, and then it broke. fixed it again now. an example of it working is for the last layer, `flurstueck`

---

*[src/app/geo/legend.service.js, line 134 at r7](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LFcfl0s9ueLsIVYaaev:-LFclvoJ2D9-veRrT9_x:b-mhsq5s) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/6dce2a4388fe7f4d2bfec5bd2ac44106b3673cc5/src/app/geo/legend.service.js#L134)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

not returning anything anymore
</blockquote></details>

donethanks

---

*[src/app/ui/loader/layer-source.service.js, line 148 at r7](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763#-LFchbt6DIe5t4Vhx30C:-LFcpekI-F9aT7_EH9hk:bh6480v) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/6dce2a4388fe7f4d2bfec5bd2ac44106b3673cc5/src/app/ui/loader/layer-source.service.js#L148)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

maybe limit = 200?  10 will cause a lot of server requests for any decent sized layer
</blockquote></details>

changed this based on our discussion

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763)*
<!-- Sent from Reviewable.io -->

james-rae commented:
Reviewed 1 of 1 files at r8, 5 of 5 files at r9.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2763)*
<!-- Sent from Reviewable.io -->
